File: java/L400/LeetCode0401.java
package L400;

import java.util.ArrayList;
import java.util.List;

public class LeetCode0401 {
    private static final int[] times = new int[]{8,4,2,1,32,16,8,4,2,1};
    public List<String> readBinaryWatch(int turnedOn) {
        List<String> res = new ArrayList<>();
        trace(turnedOn, 0, 0, 0, 0, res);
        return res;
    }

    private void trace(int turnedOn, int onCount, int index, int hour, int minute, List<String> res) {
        if (onCount == turnedOn) {
            res.add(timeToString(hour, minute));
            return;
        }
        for (int i = index;i < 10;++i) {
            if (i <= 3) {
                hour += times[i];
                if (hour <= 11) {
                    trace(turnedOn, onCount + 1, i + 1, hour, minute, res);
                }
                hour -= times[i];
            } else {
                minute += times[i];
                if (minute <= 59) {
                    trace(turnedOn, onCount + 1, i + 1, hour, minute, res);
                }
                minute -= times[i];
            }
        }
    }

    private String timeToString(int hour, int minute) {
        StringBuilder sb = new StringBuilder();
        sb.append(hour);
        sb.append(":");
        if (minute < 10) {
            sb.append('0');
        }
        sb.append(minute);
        return sb.toString();
    }

    public static void main(String[] args) {
        LeetCode0401 code = new LeetCode0401();
        code.readBinaryWatch(1);
    }
}

File: cpp/L400/LeetCode0401.cc
#include <string>
using std::string;
#include <vector>
using std::vector;

class Solution {
    constexpr static int HEAD_MASK{0b1111000000};
public:
    vector<string> readBinaryWatch(int turnedOn) {
        vector<string> r;
        for (int i = 0;i < 1024;++i) {
            int hour{0}, minute{0};
            int one{0};
            for (int j = 0;j < 6;++j) {
                if (((i >> j) & 1) == 0) {
                    minute = minute * 2;
                } else {
                    minute = minute * 2 + 1;
                    ++one;
                }
            }
            for (int j = 6;j < 10;++j) {
                if (((i >> j) & 1) == 0) {
                    hour *= 2;
                } else {
                    hour = hour * 2 + 1;
                    ++one;
                }
            }
            if (one != turnedOn || hour > 11 || minute > 59) {
                continue;
            }
            string t;
            t += std::to_string(hour) + ":";
            if (minute < 10) {
                t += "0" + std::to_string(minute);
            } else {
                t += std::to_string(minute);
            }
            r.push_back(t);
        }
        return r;
    }
};
